File: 백준/Gold/11729. 하노이 탑 이동 순서/하노이 탑 이동 순서.py
import sys
input = sys.stdin.readline
output = sys.stdout.write

def hanoi(n,fromm,to,mid):
    global answer, count
    count+= 1
    if(n == 1):
        answer.append(f"{fromm} {to}\n")
        return
    hanoi(n-1, fromm, mid, to)
    answer.append(f"{fromm} {to}\n")
    hanoi(n - 1, mid, to, fromm)

n = int(input())
answer =[]
count = 0
hanoi(n,1,3,2)
output(f"{count}\n")
output(''.join(answer))

File: 백준/Gold/11729. 하노이 탑 이동 순서/하노이 탑 이동 순서.java
import java.util.*;
import java.io.*;

public class Main {
	static int N,count;
	static StringBuilder sb = new StringBuilder();
	
	static void hanoi(int n,int from, int to, int mid) {
		count += 1;
		if(n == 1) {
			sb.append(from).append(' ').append(to).append('\n');
			return;
		}
		hanoi(n-1,from,mid,to);
		sb.append(from).append(' ').append(to).append('\n');
		hanoi(n-1,mid,to,from);
	}
	public static void main(String[] args) throws Exception{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		N = Integer.parseInt(br.readLine());
		
		hanoi(N,1,3,2);
		System.out.println(count);
		System.out.println(sb);
	}

}
